power_grid_model_c/power_grid_model/include/power_grid_model/main_core/update.hpp

```diff
@@ -94,44 +104,13 @@ inline void update_inverse(MainModelState<ComponentContainer> const& state, Forw
         },
         begin, end, sequence_idx);
 }
-
-template <symmetry_tag sym>
-inline void update_y_bus(MathState& math_state, std::vector<MathModelParam<sym>> const& math_model_params) {
-    auto& y_bus_vec = [&math_state]() -> auto& {
-        if constexpr (is_symmetric_v<sym>) {
-            return math_state.y_bus_vec_sym;
-        } else {
-            return math_state.y_bus_vec_asym;
-        }
-    }
-    ();
-
-    assert(y_bus_vec.size() == math_model_params.size());
-
-    for (Idx i = 0; i != static_cast<Idx>(y_bus_vec.size()); ++i) {
-        y_bus_vec[i].update_admittance(std::make_shared<MathModelParam<sym> const>(std::move(math_model_params[i])));
-    }
-}
-
-template <symmetry_tag sym>
-inline void update_y_bus(MathState& math_state, std::vector<MathModelParam<sym>> const& math_model_params,
-                         std::vector<MathModelParamIncrement> const& math_model_param_increments) {
-    auto& y_bus_vec = [&math_state]() -> auto& {
-        if constexpr (is_symmetric_v<sym>) {
-            return math_state.y_bus_vec_sym;
-        } else {
-            return math_state.y_bus_vec_asym;
-        }
-    }
-    ();
-
-    assert(y_bus_vec.size() == math_model_params.size());
-
-    for (Idx i = 0; i != static_cast<Idx>(y_bus_vec.size()); ++i) {
-        y_bus_vec[i].update_admittance_increment(
-            std::make_shared<MathModelParam<sym> const>(std::move(math_model_params[i])),
-            math_model_param_increments[i]);
-    }
+template <component_c Component, class ComponentContainer, std::forward_iterator ForwardIterator,
+          std::output_iterator<typename Component::UpdateType> OutputIterator>
+    requires model_component_state_c<MainModelState, ComponentContainer, Component>
+inline void update_inverse(MainModelState<ComponentContainer> const& state, ForwardIterator begin, ForwardIterator end,
+                           OutputIterator destination) {
+    return update_inverse<Component>(state, begin, end, destination,
+                                     get_component_sequence<Component>(state, begin, end));
```

Comment:
Does the length justify making it a (-n inline) function?

Reply:
length is independent of `inline`. every function has to have one of the following to correctly resolve the linking

* `extern` (maybe forward-declared)
  * this tells the compiler that this function shall be exported by the binaries.
  * `int main(argc*, argv**)` xor `int main()` are implied to be `extern`
* declared `constexpr`
  * implies `inline`
  * has the additional directive that the compiler may use the function at compile-time whenever possible.
  * this may also help with compilation speed because the code may be partially compiled even before inlining
* inlined using `inline`
  * this means that the compiler can include the code where it is used and doesn't need to link it
* be unique inside a translation unit
  * this is a little bit hard to explain but basically it amounts to e.g. a cpp file inside a nameless namespace via `namespace { /* ... */ }`)
* inside a `class` or `struct`
  * this implies `inline` if you also create the body inside the class declaration itself
  * if you do forward declare the function instead, it is added to the namespace lookup, which the linker can use at a later stage. The most common use case for this is object oriented libraries that are not header-only
    * the `typedef` definition (implied by `struct` and `class` in C++) makes sure that the function body is uniquely defined in namespace lookup
* be unique and never ever included from another location
  * different compilers may have different linking behaviour if you don't specify functions inside a nameless namespace, so it is bad practice to depend on this.
  * `doctest` abuses this possibility to register the tests to the executable, which is why test cases cannot be declared in nameless namespaces

if you do not satisfy the above (e.g. by forgetting a `constexpr` or `inline` function) and effectively end up declaring the the same function name and signature multiple times (e.g. by including the same header in 2 or more other headers) then you will get compiler errors or - even worse and more difficult to debug - linking errors

Reply:
Yet another nice read but not exactly what I was asking.
I think I get the answer to my question, which was the whole point of providing such overloads.